### PR TITLE
Fire multiple events for each activity topic on activity completion

### PR DESCRIPTION
**Pull request by cissyxyu**
## WHAT
Instead of sending one segment event on activity completion, send an event for each topic attached to that activity, with the relevant data for that topic.

## WHY
We now allow activities to have up to 3 topics, but the segment data is not capturing all 3 of an activity's topics on activity completion. After discussing various strategies with Curriculum, we've decided that the best approach would be to just send multiple segment events on activity completion, one for each of an activity's topics.

## HOW
When an activity is completed, fire an event for each of that activity's topics, with the level one, two, and three topic data for that topic.

### Screenshots
(If applicable. Also, please censor any sensitive data)

### Notion Card Links
https://www.notion.so/quill/PS-Update-how-we-send-topic-data-to-Segment-e7edc7e4a621413cb9c952825612a5ae

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  YES
Have you deployed to Staging? | Not yet - deploying now!
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | Yes

### CODE REVIEW

File: services/QuillLMS/app/services/analytics/segment_analytics.rb [hunk 1]
```diff
@@ -72,11 +72,18 @@ def track_activity_pack_assignment(teacher_id, unit_id)
...
   def track_activity_completion(user, student_id, activity, activity_session)
-    track({
-      user_id: user&.id,
-      event: SegmentIo::BackgroundEvents::ACTIVITY_COMPLETION,
-      properties: activity.segment_activity.content_params.merge({student_id: student_id})
-    })
+    activity_topics = activity.topics
+    activity_topics.each do |topic|
+      level_one_topic = topic
+      level_two_topic = Topic.find(topic.parent_id)
```
Comment (brendanshean) [inline services/QuillLMS/app/services/analytics/segment_analytics.rb]: Some refactoring suggestions for consistency and also to break up the long line:

```
  activity_topics.each do |topic_level_one|
    topic_level_two = Topic.find(topic_level_one.parent_id)
    topic_level_three = Topic.find(topic_level_two.parent_id)

    properties = { 
      student_id: student_id, 
      topic_level_one: topic_level_one.name, 
      topic_level_two: topic_level_two.name, 
      topic_level_three: topic_level_three.name, 
    }

    track({
        user_id: user&.id,
        event: SegmentIo::BackgroundEvents::ACTIVITY_COMPLETION,
        properties: activity.segment_activity.content_params.merge(properties)
    })
  ```

File: services/QuillLMS/spec/services/analytics/segment_analytics_spec.rb [hunk 2]
```diff
@@ -70,19 +70,28 @@
 
   context 'tracking activity completion' do
     let(:teacher) { create(:teacher) }
+    let(:topic) { create(:topic, level: 1) }
```
Comment (brendanshean) [inline services/QuillLMS/spec/services/analytics/segment_analytics_spec.rb]: For consistency, I would use:

```
let(:topic1) { create(:topic, level: 1) }
let(:topic2) { create(:topic, level: 1) }
```

or (less preferred):

```
let(:first_topic) { create(:topic, level: 1) }
let(:second_topic) { create(:topic, level: 1) }
```

File: services/QuillLMS/spec/services/analytics/segment_analytics_spec.rb [hunk 3]
```diff
@@ -70,19 +70,28 @@
...
+    let(:second_topic) { create(:topic, level: 1) }
     let(:activity) { create(:diagnostic_activity) }
     let(:student) { create(:student) }
     let(:activity_session) { create(:activity_session) }
 
     it 'sends an event with information about the activity' do
+      create(:activity_topic, topic: topic, activity: activity)
```
Comment (brendanshean) [inline services/QuillLMS/spec/services/analytics/segment_analytics_spec.rb]: The conventional RSpec pattern is to create/initialize everything in `let` and `before` block _before_ the `it` block where the test is run.  Therefore, move 80-81 into a before block:

```
  before do
    create(:activity_topic, topic: topic1, activity: activity)
    create(:activity_topic, topic: topic2, activity: activity)
  end
```

File: services/QuillLMS/spec/services/analytics/segment_analytics_spec.rb [hunk 4]
```diff
@@ -257,11 +267,14 @@
     end
 
     it '#track_activity_pack_completion sends the expected data' do
+      create(:activity_topic, topic: topic, activity: unit_activity2.activity)
```
Comment (brendanshean) [inline services/QuillLMS/spec/services/analytics/segment_analytics_spec.rb]: I would also move 270 into a before block:

```
before { create(:activity_topic, topic: topic, activity: unit_activity2.activity) }
```

File: services/QuillLMS/spec/services/analytics/segment_analytics_spec.rb [hunk 5]
```diff
@@ -70,19 +70,28 @@
...
       expect(track_calls[0][:event]).to eq(SegmentIo::BackgroundEvents::ACTIVITY_COMPLETION)
       expect(track_calls[0][:user_id]).to eq(teacher.id)
       expect(track_calls[0][:properties][:activity_name]).to eq(activity.name)
       expect(track_calls[0][:properties][:tool_name]).to eq('Diagnostic')
       expect(track_calls[0][:properties][:student_id]).to eq(student.id)
+      expect(track_calls[0][:properties][:topic_level_one]).to eq(topic.name)
+      expect(track_calls[0][:properties][:topic_level_two]).to eq(Topic.find(topic.parent_id).name)
+      expect(track_calls[0][:properties][:topic_level_three]).to eq(Topic.find(Topic.find(topic.parent_id).parent_id).name)
+      expect(track_calls[1][:event]).to eq(SegmentIo::BackgroundEvents::ACTIVITY_COMPLETION)
+      expect(track_calls[1][:properties][:topic_level_one]).to eq(second_topic.name)
```
Comment (brendanshean) [inline services/QuillLMS/spec/services/analytics/segment_analytics_spec.rb]: For consistency, what about 
`expect(track_calls[1][:properties][:topic_level_two])` and 
`expect(track_calls[1][:properties][:topic_level_three])`

**Review by brendanshean — APPROVED** [review]
Some minor refactoring comments, but looks good otherwise.

**Review by eadams17 — APPROVED** [review]

**Review by emilia-friedberg — APPROVED** [review]

**Pull request merged**